core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -816,4 +877,53 @@ public void stateChanged(BufferState newState)
             }
         }
     }
+
+    private static final class DriverAndTaskTerminationTracker
+            implements Runnable
+    {
+        private final TaskStateMachine taskStateMachine;
+        private final AtomicLong liveCreatedDrivers = new AtomicLong();
+
+        private DriverAndTaskTerminationTracker(TaskStateMachine taskStateMachine)
+        {
+            this.taskStateMachine = requireNonNull(taskStateMachine, "taskStateMachine is null");
+        }
+
+        @Override
+        public void run()
+        {
+            driverDestroyed();
+        }
+
+        public boolean tryCreateNewDriver()
+        {
+            // Eagerly increment the counter before checking the state machine
+            liveCreatedDrivers.incrementAndGet();
+            // If termination has started already, we need to decrement the counter and check for termination complete
+            if (taskStateMachine.getState().isTerminatingOrDone()) {
+                driverDestroyed();
+                return false;
+            }
+            return true;
+        }
+
+        public void driverDestroyed()
+        {
+            if (liveCreatedDrivers.decrementAndGet() == 0) {
+                checkTaskTermination();
+            }
+        }
+
+        public long getLiveCreatedDrivers()
+        {
+            return liveCreatedDrivers.get();
+        }
+
+        public void checkTaskTermination()
+        {
+            if (liveCreatedDrivers.get() == 0 && taskStateMachine.getState().isTerminating()) {
```

Comment:
With bugs this will go negative and the termination will never be called.  With the race in `tryCreateNewDriver` this could be positive when this is checked

Reply:
In the event of a race in `tryCreateNewDriver()` this could go positive for any particular check, but in that case the thread causing the race in `tryCreateNewDriver()` will notice the state machine is terminating and decrement the count back down and then re-check for termination.

I agree that if there's a bug and this goes negative then we're stuck but... what would we propose to do instead? Hard-fail the task state machine immediately even though drivers are still potentially running? Just consider anything negative as "successfully terminated"? I'm open to suggestions about what the best option is here.

Reply:
Here is what I was thinking.  I don't know if this actually happens, but I think the code allows for this.
```
thread 1: tryCreateNewDriver incrementAndGet
thread 1: if isTerminatingOrDone is not entered
thread x: terminate task
thread 2: liveCreatedDrivers.get() will equal 1 and state is terminating is true, so terminationComplete is not called

Reply:
For safety, I typically check `<= 0`, it mask over bugs (you could log), but it prevents bugs from locking up resources.  Also, calling check termination periodically after termination until the task is actually all cleaned up.  Again this can mask bugs, but it is better than locking up a server.

Reply:
After off line discussion, my example above is covered by the call back handler added to the created driver.